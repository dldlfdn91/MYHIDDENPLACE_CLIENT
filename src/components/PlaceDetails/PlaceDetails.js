import React, { Component } from "react";
import { NavLink } from "react-router-dom";
import ReactMarkdown from "react-markdown";
import Rating from "react-rating";
import PlaceMap from "../Map/PlaceMap";
import {
  MdStar,
  MdStarBorder,
  MdFavorite,
  MdFavoriteBorder,
  MdLocationOn,
  MdPerson,
  MdPoll,
  MdDateRange
} from "react-icons/md";
import "./PlaceDetails.scss";

class PlaceDetails extends Component {
  constructor(props) {
    super(props);

    this.state = {
      commentText: "",
      score: 0
    };

    this.handleCommentChange = this.handleCommentChange.bind(this);
    this.handleScoreClick = this.handleScoreClick.bind(this);
    this.submitComment = this.submitComment.bind(this);
    this.submitFavoritePlace = this.submitFavoritePlace.bind(this);
    this.removeFavoritePlace = this.removeFavoritePlace.bind(this);
  }

  componentDidMount() {
    const placeId = this.props.match.params.place_id;
    this.props.onLoad(placeId);
    this.props.onLoadComment(placeId);
  }

  componentDidUpdate() {
    const placeId = this.props.match.params.place_id;
    this.props.onLoadComment(placeId);
  }

  handleCommentChange = event => {
    this.setState({
      commentText: event.target.value
    });
  };

  handleScoreClick = rate => {
    this.setState({
      score: rate
    });
  };

  submitComment = event => {
    event.preventDefault();

    const placeId = this.props.match.params.place_id;
    const text = this.state.commentText;
    const score = this.state.score;

    this.props.onSubmit(placeId, text, score);

    this.setState({
      commentText: "",
      score: 0
    });
  };

  submitFavoritePlace = event => {
    event.preventDefault();

    const placeId = this.props.match.params.place_id;
    this.props.registerFavoritePlace(placeId);
  };

  removeFavoritePlace = event => {
    event.preventDefault();

    const placeId = this.props.match.params.place_id;
    this.props.deleteFavoritePlace(placeId);
  };

  render() {
    return (
      <div className="PlaceDetails">
        <div className="place-details-container">
          <div className="place-details-header">
            <div className="place-details-header-info">
              <h1 className="place-details-title">
                {this.props.placeDetails.title}
              </h1>
              <span className="place-details-location">
                <MdLocationOn className="MdLocationOn" />
                {this.props.placeDetails.address}
              </span>
              <span className="place-details-author">
                <MdPerson className="MdPerson" />
                {this.props.placeDetails.created_by}
              </span>
              <span className="place-details-rate">
                {this.props.comments.length === 0 ? (
                  <span className="place-details-no-rate">
                    <MdPoll className="MdPoll" />
                    평점이 없습니다 -
                  </span>
                ) : (
                  <Rating
                    emptySymbol={<MdStarBorder className="MdStarBorder" />}
                    fullSymbol={<MdStar className="MdStar" />}
                    initialRating={this.props.avgScore}
                    className="place-details-rate-image"
                    readonly
                  />
                )}
              </span>
              <span className="place-details-score">
                {this.props.avgScore}점
              </span>
            </div>

            <div className="place-favorite">
              {this.props.favoritePlace.length !== 0 ? (
                this.props.favoritePlace.find(
                  place => place === this.props.placeDetails._id
                ) ? (
                  <MdFavorite
                    onClick={this.removeFavoritePlace}
                    className="MdFavorite"
                  />
                ) : (
                  <MdFavoriteBorder
                    onClick={this.submitFavoritePlace}
                    className="MdFavoriteBorder"
                  />
                )
              ) : (
                <MdFavoriteBorder onClick={this.submitFavoritePlace} />
              )}
            </div>
          </div>

          <div className="place-details-main">
            <div className="place-details-main-left-container">
              <div className="place-details-picture">
                <img
                  src={this.props.placeDetails.place_picture}
                  alt="place detail pic"
                />
              </div>

              <div className="place-details-map">
                <PlaceMap
                  location={this.props.placeDetails.location}
                  address={this.props.placeDetails.address}
                />
              </div>
            </div>

            <div className="place-details-main-right-container">
              <div className="place-details-descriptoin">
                <ReactMarkdown source={this.props.placeDetails.description} />
              </div>

              <div className="place-details-tags">
                {this.props.placeDetails.tag
                  ? this.props.placeDetails.tag.map((tag, index) => (
                      <li key={index}>
                        <span>#{tag}</span>
                      </li>
                    ))
                  : ""}
              </div>

              <div className="place-details-comments">
                {this.props.isAuthenticated ? (
                  <div className="write-comments">
                    <form>
                      <div className="write-comments-rate">
                        <Rating
                          initialRating={this.state.score}
                          emptySymbol={
                            <MdStarBorder className="MdStarBorder" />
                          }
                          fullSymbol={<MdStar className="MdStar" />}
                          fractions={2}
                          className="place-details-rate-image"
                          onClick={rate => this.handleScoreClick(rate)}
                        />
                      </div>
                      <input
                        type="text"
                        name="content"
                        value={this.state.commentText}
                        className="write-comments-text"
                        onChange={event => this.handleCommentChange(event)}
                        placeholder="저작권 등 다른 사람의 권리를 침해하거나 명예를 훼손하는 게시물은 제재를 받을 수 있습니다."
                      />
                      <input
                        type="submit"
                        className="write-comments-submit"
                        onClick={event => this.submitComment(event)}
                      />
                    </form>
                  </div>
                ) : (
                  <div className="require-login">
                    <NavLink exact to={"/signin"}>
                      <button>인</button>
                    </NavLink>
                    <span> 후 댓글을 작성할 수 있습니다.</span>
                  </div>
                )}
                <div className="show-comments">
                  <ul>
                    {this.props.comments.length !== 0 ? (
                      this.props.comments.map(comment => (
                        <li key={comment._id} className="comments-list">
                          <div className="comment-by">
                            <MdPerson className="MdPerson" />
                            {comment.created_by}
                          </div>
                          <div className="comment-text">
                            {comment.comment_text}
                          </div>
                          <div className="comment-date">
                            <MdDateRange className="MdDateRange" />
                            {comment.created_at.slice(0, 10)}
                          </div>
                        </li>
                      ))
                    ) : (
                      <div className="comment-error-msg">
                        {this.props.commentErrorMsg}
                      </div>
                    )}
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default PlaceDetails;
