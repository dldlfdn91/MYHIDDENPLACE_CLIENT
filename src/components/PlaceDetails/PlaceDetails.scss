$mainColor: #1b5ac2;
$subColor: #0c2236;
$type1: #afb1be;
$type2: #e9ecef;
$type3: #f03e3e;
$type4: #fab005;
$font: "Noto Sans KR", sans-serif;

.PlaceDetails {
  font-family: $font;
  .place-details-container {
    width: 65%;
    margin: 0 auto;
    padding-top: 8%;
    .place-details-header {
      position: relative;
      margin-bottom: 5%;
      .place-details-header-info {
        padding: 20px 20px 20px 0;
        text-align: left;
        .place-details-title {
          margin: 0;
          margin-bottom: 10px;
          font-size: 35px;
        }
        .place-details-location {
          margin-right: 10px;
          color: $type1;
          font-size: 15px;
        }
        .place-details-author {
          margin-right: 10px;
          color: $type1;
          font-size: 15px;
        }
        .place-details-rate {
          margin-right: 10px;
          color: $type1;
          .place-details-no-rate {
            font-size: 15px;
          }
          .place-details-rate-image {
            color: $type1;
            font-size: 15px;
            .MdStarBorder {
              color: $type4;
            }
            .MdStar {
              color: $type4;
            }
          }
        }
        .place-details-score {
          color: $type1;
          font-size: 15px;
        }
      }
      .place-favorite {
        position: absolute;
        top: 40px;
        right: 30px;
        font-size: 40px;
        cursor: pointer;
        .MdFavorite {
          color: $type3;
        }
        .MdFavoriteBorder {
          color: $type3;
        }
      }
    }

    .place-details-main {
      display: flex;
      .place-details-main-left-container {
        width: 50%;
        height: 848px;
        margin-top: 25px;
        .place-details-picture {
          img {
            width: 100%;
            height: 340px;
            border-radius: 5px;
          }
        }
      }
      .place-details-main-right-container {
        width: 50%;
        padding-left: 30px;
        text-align: left;
        .place-details-description {
          margin-bottom: 30px;
        }
        .place-details-tags {
          margin-bottom: 5px;
          li {
            display: inline-block;
            margin-right: 5px;
            color: $mainColor;
          }
        }
        .place-details-comments {
          .write-comments {
            position: relative;
            width: 90%;
            height: 102px;
            margin: 35px 0;
            padding: 10px 0 0 10px;
            border: 1px solid $type1;
            .write-comments-rate {
              .place-details-rate-image {
                color: $type1;
                font-size: 20px;
                .MdStarBorder {
                  color: $type4;
                }
                .MdStar {
                  color: $type4;
                }
              }
            }
            .write-comments-text {
              display: block;
              width: 100%;
              margin-top: 5px;
              outline: none;
              font-size: 15px;
              text-align: left;
              outline: 0;
              border: 0;
              border-bottom: 1px solid $type1;
            }
            .write-comments-submit {
              position: absolute;
              top: 64px;
              right: -1px;
              padding: 10px 20px;
              background: $mainColor;
              color: #ffffff;
              font-size: 13px;
              font-weight: bold;
              border: none;
            }
          }
          .require-login {
            margin: 50px 0;
            text-align: center;
            button {
              padding: 0;
              border: none;
              font-size: 16px;
              color: $mainColor;
              cursor: pointer;
              &:hover {
                text-decoration: underline;
              }
            }
          }
          .show-comments {
            ul {
              margin: 0;
              padding: 0;
              .comments-list {
                width: 90%;
                text-align: left;
                margin-bottom: 20px;
                border-bottom: 3px double $type1;
                .comment-by {
                  margin-bottom: 2px;
                  font-size: 15px;
                  color: $type1;
                  .MdPerson {
                    margin-right: 5px;
                  }
                }
                .comment-text {
                  margin-bottom: 2px;
                  font-size: 18px;
                  color: $subColor;
                }
                .comment-date {
                  color: $type1;
                  .MdDateRange {
                    margin-right: 5px;
                  }
                }
                .comment-error-msg {
                }
              }
            }
          }
        }
      }
    }
  }
}
