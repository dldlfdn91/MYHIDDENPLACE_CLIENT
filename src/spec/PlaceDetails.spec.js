import React from "react";
import { shallow } from "enzyme";
import PlaceDetails from "../components/CreatePlace/CreatePlace";

describe("<PlaceDetails />", () => {
  let wrapper;
  const props = {
    onLoad: jest.fn(),
    placeDetails: [
      {
        _id: "5daf9244d360a257663569a5",
        place_picture: [
          "https://my-hidden-place.s3.ap-northeast-2.amazonaws.com/1571787331747.jpg"
        ],
        tag: ["서울", "런던", "파리"],
        title: "도시와 강",
        created_by: "test",
        created_at: "2019-10-22T23:35:32.226+00:00",
        description: "도시와 강",
        address: "서울",
        location: {
          coordinates: ["-0.08769340000003467", "51.5082192"]
        }
      }
    ],
    isMarkerShown: true,
    onSubmit: jest.fn(),
    isAuthenticated: true,
    onLoadComment: jest.fn(),
    comment: [],
    commentErrorMsg: "댓글이 없습니다.",
    avgScore: 0,
    registerFavoritePlace: jest.fn(),
    favoritePlace: jest.fn(),
    deleteFavoritePlace: jest.fn()
  };

  beforeEach(() => {
    wrapper = shallow(<PlaceDetails {...props} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it.skip("must have title", function() {
    expect(
      wrapper.contains(<h1 className="place-details-title">도시와 강</h1>)
    ).toBe(true);
  });
});
