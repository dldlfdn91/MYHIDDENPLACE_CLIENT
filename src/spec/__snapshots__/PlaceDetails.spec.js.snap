// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaceDetails /> matches snapshot 1`] = `ShallowWrapper {}`;
